### PR TITLE
Autocmds configuration

### PR DESCRIPTION
**Pull request by jalil-salame**
A minimum viable product for augroups and autocmds, lacks testing but the lua code looks fine, there are some things which don't work/I don't know how to implement (see TODOs):

- Assertions
- Setting a default value from the attribute name

I'll test this tomorrow and see if anything needs fixing, but it's probably minor so you can start reviewing it now.

EDIT: It has been tested and works flawlessly, I still have some pet peeves though:

- [ ] No assertions (I don't know where to put the `assertions = [];`)
- [ ] I can't figure out how to set the name of the augroups (I'd like `augroups.hightlighOnYank = {...};` instead of `augroups = [{name = "highlightOnYank"; ...}];`)

### CODE REVIEW

File: helper/augroups.nix [hunk 1]
```diff
@@ -0,0 +1,189 @@
...
+
+  inherit (lib) mkOption;
+  inherit (lib.strings) concatMapStringsSep;
+  inherit (lib.attrsets) mapAttrsToList;
+  inherit (lib.types) either bool str listOf nullOr submodule lines int;
+
+  autocmdOpts = {
+    options = {
+      event = mkOption {
+        # TODO: make enum of nvim events
```
Comment (jooooscha) [inline helper/augroups.nix:15]: Could you add this before merging?

File: nixneovim.nix [hunk 2]
```diff
@@ -178,6 +180,30 @@ in
           };
         '';
       };
+
+      augroups = mkOption {
+        default = { };
+        type = let inherit (augroups) augroupOptions; in types.attrsOf augroupOptions;
```
Comment (jooooscha) [inline nixneovim.nix:186]: Why did you make this so complicated and long?

Could you change it to `types.attrsOf augroups.augroupsOptions`?

File: helper/augroups.nix [hunk 3]
```diff
@@ -0,0 +1,189 @@
...
+               };
+             in ''
+    do
+      local events = ${events}
+      local opts = ${toLuaObject opts}
+      vim.api.nvim_create_autocmd(events, opts)
+    end
+  '';
+
+  genAugroup = {name, autocmds, clear}: let
```
Comment (jooooscha) [inline helper/augroups.nix:144]: Could you add brief documentations for this function?

Also for genAugroups and genAutocmd

File: nixneovim.nix [hunk 4]
```diff
@@ -178,6 +180,30 @@ in
...
+        description = ''
+          Custom autocmd groups
+        '';
+        example = ''
+          augroups.highlightOnYank = {
+            autocmds = [{
+              event = "TextYankPost";
+              pattern = "*";
+              luaCallbak = ''\''
```
Comment (jooooscha) [inline nixneovim.nix:195]: Small typo. Should be `luaCallback`

File: helper/augroups.nix [hunk 5]
```diff
@@ -0,0 +1,189 @@
...
+      };
+      autocmds = mkOption {
+        type = listOf (submodule autocmdOpts);
+        description = ''
+          The autocmds that are part of this augroup.
+
+          See :help nvim_create_autocmd()
+        '';
+      };
+      clear = mkOption {
```
Comment (jooooscha) [inline helper/augroups.nix:178]: In `helper/custom_options.nix` I have defined several helper functions, which make writing simple bool options much shorter. Could you make use of them here?

**Review by jooooscha — CHANGES_REQUESTED** [review]
Looks good. Thank you for your work.

I have only smaller remarks. Otherwise, I think this can be merged

File: nixneovim.nix (hunk 2 above)
Comment (jalil-salame) [reply nixneovim.nix:186]: First time actually writing so much nix code, I'm busy this weekend so you can expect this to be cleaned up and done by next weekend

**Comment by jooooscha** [comment]
Looks good. Will do some testing locally, but should be ready for merging then 👍
I will tackle your todos after that.

Thanks for your work and fast response.

**Comment by jooooscha** [comment]
Because of #69, I ran `nix flake check -L` on your branch and got the following error in `augroups.nix`:

```
❯ nix flake check -L
error: value is a function while a set was expected

       at /nix/store/xk2r4ys4jkwlyi5p2jjmjam0anmzmc4n-source/helper/augroups.nix:4:7:

            3|   toLuaObject,
            4| }: let
             |       ^
            5|   inherit (import ../lib.nix) rawLua;
(use '--show-trace' to show detailed location information)
```

**Comment by jalil-salame** [comment]
I can't reproduce the error, I seem to get the same, value is a function, but for the docs, I might need to bisect the PR or rework it from scratch...

**Comment by jooooscha** [comment]
Weird. Have you committed and pushed everything?

**Comment by jalil-salame** [comment]
AFAIK yes, I'm on NixOS unstable, maybe having a different version of nix is messing stuff up?

I'll check what version I'm running and try updating it

**Comment by jalil-salame** [comment]
Bisecting the commits, show that I broke something while trying to address the review comments, sadly I don't have time to investigate, I will probably come back to this within 2 weeks.

**Comment by jalil-salame** [comment]
Okay, so... It seems that `boolOption` from `helpers/custom_options.nix` was not being fully appplied??? Like, it was still a function after passing the arguments, so I just wrote my own version at the top of the file and that fixed it, but we should probably look into what was happening.

https://github.com/NixNeovim/NixNeovim/pull/66/files#diff-1797869db97503d2a3eb55600f343fec629e2879ead746468969c2a32c502908R12-R16

I thought I was doing everything properly

**Comment by jalil-salame** [comment]
I fixed it, I was just being dumb

**Comment by jooooscha** [comment]
Looks good. Thanks for all your work.

Last thing missing is my opinion would be a short example config. Maybe you could add something to the readme?

**Comment by jalil-salame** [comment]
All fingers crossed, but I think I am done!

**Comment by jooooscha** [comment]
Looks great. Tests only fail in Action; locally, everything is fine.

Thanks for your work and fast responses. 👍

**Pull request merged**